src/nbl/ext/ImGui/ImGui.cpp

```diff
@@ -304,287 +451,468 @@ namespace nbl::ext::imgui
 		}
+
+		auto buffer = m_mdi.buffer->getBuffer();
+		auto binding = buffer->getBoundMemory();
+
+		const auto validation = std::to_array
+		({
+			std::make_pair(buffer->getCreationParams().usage.hasFlags(MDI::MDI_BUFFER_REQUIRED_USAGE_FLAGS), "MDI buffer must be created with IBuffer::EUF_INDIRECT_BUFFER_BIT | IBuffer::EUF_INDEX_BUFFER_BIT | IBuffer::EUF_VERTEX_BUFFER_BIT | IBuffer::EUF_SHADER_DEVICE_ADDRESS_BIT enabled!"),
+			std::make_pair(bool(buffer->getMemoryReqs().memoryTypeBits & m_creationParams.utilities->getLogicalDevice()->getPhysicalDevice()->getUpStreamingMemoryTypeBits()), "MDI buffer must have up-streaming memory type bits enabled!"),
+			std::make_pair(binding.memory->getAllocateFlags().hasFlags(MDI::MDI_BUFFER_REQUIRED_ALLOCATE_FLAGS), "MDI buffer's memory must be allocated with IDeviceMemoryAllocation::EMAF_DEVICE_ADDRESS_BIT enabled!"),
+			std::make_pair(binding.memory->isCurrentlyMapped(), "MDI buffer's memory must be mapped!"), // streaming buffer contructor already validates it, but cannot assume user won't unmap its own buffer for some reason (sorry if you have just hit it)
+			std::make_pair(binding.memory->getCurrentMappingAccess().hasFlags(getRequiredAccessFlags(binding.memory->getMemoryPropertyFlags())), "MDI buffer's memory current mapping access flags don't meet requirements!")
+		});
+
+		for (const auto& [ok, error] : validation)
+			if (!ok)
+			{
+				m_creationParams.utilities->getLogger()->log(error, system::ILogger::ELL_ERROR);
+				assert(false);
+			}
 	}
 
-	//-------------------------------------------------------------------------------------------------
-	// TODO: Handle mouse cursor for nabla
-	//static void UpdateMouseCursor()
-	//{
-		//auto & io = ImGui::GetIO();
-
-		//if (io.ConfigFlags & ImGuiConfigFlags_NoMouseCursorChange)
-		//{
-		//    return;
-		//}
-		//ImGuiMouseCursor imgui_cursor = ImGui::GetMouseCursor();
-		//if (io.MouseDrawCursor || imgui_cursor == ImGuiMouseCursor_None)
-		//{
-		//    // Hide OS mouse cursor if imgui is drawing it or if it wants no cursor
-		//    MSDL::SDL_ShowCursor(MSDL::SDL_FALSE);
-		//}
-		//else
-		//{
-		//    // Show OS mouse cursor
-		//    MSDL::SDL_SetCursor(mouseCursors[imgui_cursor] ? mouseCursors[imgui_cursor] : mouseCursors[ImGuiMouseCursor_Arrow]);
-		//    MSDL::SDL_ShowCursor(MSDL::SDL_TRUE);
-		//}
-	//}
-
-	bool UI::Render(IGPUCommandBuffer* commandBuffer, int const frameIndex)
+	bool UI::render(nbl::video::IGPUCommandBuffer* const commandBuffer, nbl::video::ISemaphore::SWaitInfo waitInfo, const std::span<const VkRect2D> scissors)
 	{
+		if (!commandBuffer)
+		{
+			m_creationParams.utilities->getLogger()->log("Invalid command buffer!", system::ILogger::ELL_ERROR);
+			return false;
+		}
+
+		if (commandBuffer->getState() != IGPUCommandBuffer::STATE::RECORDING)
```

Comment:
you can also validate that a renderpass instance is open

Reply:
resolved